### PR TITLE
Add `params` and `rng` argument to all `FuncEnv` member functions

### PR DESCRIPTION
**Pull request by Kallinteris-Andreas**
# Description

Fixes https://github.com/Farama-Foundation/Gymnasium/issues/851, https://github.com/Farama-Foundation/Gymnasium/issues/861

## Type of change

- [ ] Documentation only change (no code changed)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

# Checklist:
- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/envs/functional/test_core.py [hunk 1]
```diff
@@ -12,23 +12,25 @@ def __init__(self, options: Optional[Dict[str, Any]] = None):
...
-    def terminal(self, state: np.ndarray) -> bool:
+    def terminal(self, state: np.ndarray, rng: Any) -> bool:
         return state[1] > 0
 
 
 def test_api():
     env = BasicTestEnv()
     state = env.initial(None)
-    obs = env.observation(state)
+    obs = env.observation(state, None)
```
Comment (RedTachyon) [inline tests/envs/functional/test_core.py:33]: Can we make the rng argument optional the same way as params?

**Review by RedTachyon — CHANGES_REQUESTED** [review]
At a glance seems alright, except it would be nice if the `rng` argument was also optional. Can you check if that works well, and if so, add tests for all variants of calling the functions? (implicit None, explicit None, explicit non-trivial arguments)

File: tests/envs/functional/test_core.py (hunk 1 above)
Comment (Kallinteris-Andreas) [reply tests/envs/functional/test_core.py:33]: How should `rng` default be handled?
If an environment's needs `transition` member function requires `rng`, what should be done?

In the case of parameters, there is clearly a default

Comment (Kallinteris-Andreas) [reply tests/envs/functional/test_core.py:33]: @RedTachyon

**Comment by Kallinteris-Andreas** [comment]
@RedTachyon check the abobe comment Thanks

**Comment by Kallinteris-Andreas** [comment]
@RedTachyon

**Comment by RedTachyon** [comment]
Sorry about the delay. I was a bit concerned about potential performance implications of this, since jax jit can be a bit moody sometimes with its optimizations, but from a quick test it seems to not cause any issues. We might still have to revisit performance when it comes up, but it's not a priority at the moment, and we'll see how big of a deal it is in "real" environments

So, uh, LGTM

**Review by RedTachyon — APPROVED** [review]
